### ISSUE TITLE
PHP 8.1 deprecation notices

### ISSUE THREAD

**Issue by alexgit2k** [phpseclib/phpseclib#1770]
With PHP 8.1 I get following deprecation notices with phpseclib:

**1.0**
- `phpseclib/Crypt/Base.php, line 1959: substr(): Passing null to parameter #1 ($string) of type string is deprecated`

**2.0**
- `phpseclib/Crypt/Base.php, line 2031: substr(): Passing null to parameter #1 ($string) of type string is deprecated`

**3.0**
- `phpseclib/Math/Common/FiniteField/Integer.php, line 34: Return type of phpseclib3\Math\Common\FiniteField\Integer::jsonSerialize() should either be compatible with JsonSerializable::jsonSerialize(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice`
- `phpseclib/Math/BigInteger/Engines/Engine.php, line 379: Return type of phpseclib3\Math\BigInteger\Engines\Engine::jsonSerialize() should either be compatible with JsonSerializable::jsonSerialize(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice`
- `phpseclib/Math/BigInteger.php, line 446: Return type of phpseclib3\Math\BigInteger::jsonSerialize() should either be compatible with JsonSerializable::jsonSerialize(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice`

**Comment by terrafrost** [comment]
I've fixed the 1.0 and 2.0 deprecation notices. I'd add a unit test but tbh the unit tests that exist should already be catching that. That they're not is...  strange. Like I'm able to reproduce it thusly:

```php
$a = new Crypt_RC4;
$a->setKey('aaaaaa');

echo bin2hex($a->encrypt('asdfasdf'));
```
For good measure I added that as a unit test at https://github.com/terrafrost/phpseclib/commit/1409615075cb80115c5e01bb41ddde978b147c87 and it's still passing on Travis CI on 8.1...

That said, the test I added shouldn't need to be added, anyway. I just did it to more effectively demo the fact that PHPUnit isn't failing when it encounters deprecation notices.
